cli-plugins/manager/manager.go

```diff
@@ -0,0 +1,190 @@
+		if runtime.GOOS == "windows" {
+			exe := ".exe"
+			if !strings.HasSuffix(name, exe) {
+				continue
+			}
+			name = strings.TrimSuffix(name, exe)
+		}
+		res[name] = append(res[name], filepath.Join(d, dentry.Name()))
+	}
+	return nil
+}
+
+// listPluginCandidates allows the dirs to be specified for testing purposes.
+func listPluginCandidates(dirs []string) (map[string][]string, error) {
+	result := make(map[string][]string)
+	for _, d := range dirs {
+		if err := addPluginCandidatesFromDir(result, d); err != nil {
+			return nil, err // Or return partial result?
+		}
+	}
+	return result, nil
+}
+
+// ListPluginCandidates returns a map from plugin name to the list of (unvalidated) Candidates. The list is in descending order of priority.
+func ListPluginCandidates(dockerCli command.Cli) (map[string][]string, error) {
+	return listPluginCandidates(getPluginDirs(dockerCli))
+}
+
+// ListPlugins produces a list of the plugins available on the system
+func ListPlugins(dockerCli command.Cli, rootcmd *cobra.Command) ([]Plugin, error) {
+	candidates, err := ListPluginCandidates(dockerCli)
+	if err != nil {
+		return nil, err
+	}
+
+	var plugins []Plugin
+	for _, paths := range candidates {
+		if len(paths) == 0 {
+			continue
+		}
+		c := &candidate{paths[0]}
+		p, err := NewPlugin(c, rootcmd)
+		if err != nil {
+			return nil, err
+		}
+		p.ShadowedPaths = paths[1:]
+		plugins = append(plugins, p)
+	}
+
+	return plugins, nil
+}
+
+// FindPlugin finds a valid plugin, if the first candidate is invalid then returns an error
+func FindPlugin(dockerCli command.Cli, name string, rootcmd *cobra.Command, includeShadowed bool) (Plugin, error) {
+	if !pluginNameRe.MatchString(name) {
+		// We treat this as "not found" so that callers will
+		// fallback to their "invalid" command path.
+		return Plugin{}, ErrPluginNotFound(name)
+	}
+	exename := NamePrefix + name
+	if runtime.GOOS == "windows" {
+		exename = exename + ".exe"
+	}
+	var plugin Plugin
+	for _, d := range getPluginDirs(dockerCli) {
+		path := filepath.Join(d, exename)
+
+		// We stat here rather than letting the exec tell us
+		// ENOENT because the latter does not distinguish a
+		// file not existing from its dynamic loader or one of
+		// its libraries not existing.
+		if _, err := os.Stat(path); os.IsNotExist(err) {
+			continue
+		}
+
+		if plugin.Path == "" {
+			c := &candidate{path: path}
+			var err error
+			if plugin, err = NewPlugin(c, rootcmd); err != nil {
+				return Plugin{}, err
+			}
+			if !includeShadowed {
+				return plugin, nil
+			}
+		} else {
+			plugin.ShadowedPaths = append(plugin.ShadowedPaths, path)
+		}
+	}
+	if plugin.Path == "" {
+		return Plugin{}, ErrPluginNotFound(name)
+	}
+	return plugin, nil
+}
+
+func runPluginCommand(dockerCli command.Cli, name string, rootcmd *cobra.Command, args []string) (*exec.Cmd, error) {
+	plugin, err := FindPlugin(dockerCli, name, rootcmd, false)
+	if err != nil {
+		return nil, err
+	}
+	if plugin.Err != nil {
```

Comment:
This can be moved in `FindPlugin` no?

Reply:
Yes, now that `FindPlugin` only has a single caller I think the two can be collapsed.

Reply:
Furthermore, there are no longer any callers of `findPlugin` which pass `includeShadowed=true` which leads to a whole bunch of simplification and I think means `PluginRunCommand`→`runPluginCmmand`→`findPlugin` will sanely collapse into one function. I'll try that now.